### ISSUE TITLE
Provide a way to tag the origin of an object and detect similarity

### ISSUE THREAD

**Issue by Frooxius** [Frooxius/NeosPublic#105]
Based on discussion with Outsider and Nexulan, a way to tag an object to determine its origin (world or user who tagged it) or uniqueness, that could be verifiable by anyone else.

One thing that  could be also potentially interesting is hashing and potentially asymmetric encryption on top of that (to prove it comes from a particular user). 

Could iterate over all the properties and hash them in some predictable way. Most value fields would be straightfoward by themselves, but would need a deterministic way of hashing references, since these vary depending on the instance and order in which things were loaded/constructed. 

And another issue is that some fields and parts of the structure will likely vary anyway, just by the function of the object. Could do something more like a proximity hash instead (so it would report that the object is 1 % different from the originally tagged version) or find a way to automatically exclude these (like watching it for some time in operation and including only fields that don't change).

This would for example let anyone verify that a particular object comes from a trusted source or allow to find copies - they would end up with same/similar hash for the object itself, which would already be registered with an earlier date, which would mean that somebody else constructed it first. Could have implications for the market a well.

This isn't a priority right now and requires more thought, but this is just a basic sketch of the idea.

**Comment by Frooxius** [comment]
Yeah, blockchain could be used as the decentralized verification of the original "authorship" of the object.

I think the most difficult part is going to be the hashing itself though - how to tell that the two objects are the same. Some fields will constantly change just by the operation of the object. And it could be worked around by changing a few things (like changing a floating point value by tiny amount or texture pixel by a little bit).

If it's too sensitive, it'll be easy to modify the object and have it considered to be new and unique, if it's too vague, it might end up with a lot of false duplicates.

**Comment by sirkitree** [comment]
Hrm, yeah - Decentraland must have run into this at some point. I wonder if @bnolan might have any insight or advice to lend us there?

**Comment by Frooxius** [comment]
I'm not sure they would have, depending on how they work at low level. Usually these applications manage objects as higher level things/entities, so I'm afraid this issue might be a bit more specific to the architecture of Neos itself.

In Neos everything in the world is part of one data hierarchy, with chunks of the hierarchy being replicated when you copy or transfer the objects. From it's perspective it's not clear where object really starts.

I think that enforcing something like this might be a bit of a wild goose chase. On the technical side, without serious restrictions on the freedom, it might always be trivial to circumvent the detection. On practical I'm not sure people would think to verify and challenge ownership of every little thing they see.
